### PR TITLE
improvements in find_dividers function (#20)

### PR DESCRIPTION
**Pull request by lomereiter**
Summary of changes:

* path links are represented as a numpy array throughout, of shape (number of links) x 2;
* each case in the original loop moved into its own function in utils.py; there are basic unit tests, too;
* each function operates on the whole array of path links and returns a boolean mask;
* in `utils.path_dividers` vectorized binary search does all the job (`np.searchsorted`)
* `find_dividers` now returns a pandas dataframe (from, to, path) instead of two nested sets;
* the sets are still reconstructed in `segment_matrix`, though (five lines of code)
* I discarded `discard_useless_links` function as it's dead code

### CODE REVIEW

**Comment by josiahseaman** [comment]
It's really encouraging to see progress here.  I heard we're seeing a 30x speedup in load times.  Just ping me when this is ready for review.  I've got a whole dashboard of notifications, so I don't always notice right away.  I wrote component_segmentation, so I know it well.

**Comment by lomereiter** [comment]
OK, this is ready for review. I'll create a separate issue for `segment_matrix` overhaul.

File: matrixcomponent/JSONparser.py [hunk 1]
```diff
@@ -3,18 +3,21 @@
...
 def parse(file):
-        data = []
-        with open(file) as f:
-            for line in f:
-                data.append(json.loads(line))
-
-        paths = []
-        pangenome_length = 0
-        for path in data:
+    data = []
```
Comment (subwaystation) [inline matrixcomponent/JSONparser.py]: This line is not needed anymore, because we don't use `data` anymore.

File: matrixcomponent/tests.py [hunk 2]
```diff
@@ -1,8 +1,42 @@
+import pytest
```
Comment (subwaystation) [inline matrixcomponent/tests.py]: Running `pytest matrixcomponent/tests.py` this line is not needed.

File: matrixcomponent/tests.py [hunk 3]
```diff
@@ -1,8 +1,42 @@
...
+import numpy as np
+
 from matrixcomponent.matrix import Path
```
Comment (subwaystation) [inline matrixcomponent/tests.py]: Running `pytest matrixcomponent/tests.py` this line is not needed.

**Review by subwaystation — APPROVED** [review]
For master and this branch, I get the same output. Nice refactoring @lomereiter !
I am happy, that we have at least some tests, now.
As this changes some code I would wait on @josiahseaman approval, before merging this in.

File: segmentation.py [hunk 4]
```diff
@@ -147,69 +157,76 @@ def add_adjacent_connector_column(component, next_component, schematic):
...
+        bin_ids.sort()
+
+        if bin_ids.size > 0:
+            max_bin = max(max_bin, int(bin_ids[-1]))
+
+        # we don't want these to become dividers
+        boundary_mask = utils.path_boundaries(links)
+        self_loops_mask = utils.self_loops(links)
+
+        if np.any(self_loops_mask):
```
Comment (josiahseaman) [inline segmentation.py:178]: Based on my recollection, self_loops turned out to not exist in odgi output.  I was testing for them, but they don't happen so you could save some time and complexity by removing the concept.  (That can be in a later PR)

File: segmentation.py [hunk 5]
```diff
@@ -147,69 +157,76 @@ def add_adjacent_connector_column(component, next_component, schematic):
...
+    df.drop_duplicates(inplace=True)
+
+    # all start positions of components
+    dividers = set(df["from"] + 1) | set(df["to"])
+    dividers.add(1)
+
     print(f"Largest bin_id was {max_bin}\n"
           f"Found {len(dividers)} dividers.")
     dividers.add(max_bin + 1)  # end of pangenome
-    print(f"Eliminated {len(copy_arrivals)} self-loops")
```
Comment (josiahseaman) [inline segmentation.py:193]: This has always returned zero as far as I can recall.

**Review by josiahseaman — APPROVED** [review]
I've verified all the outputs are equivalent and the code is cleaner and 30x faster.  No downside here.

File: segmentation.py (hunk 4 above)
Comment (lomereiter) [reply segmentation.py:178]: Thanks, good to know.

**Pull request merged**